File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        
        int start = 0, maxL = 0;
        
        HashMap<Character, Integer> map = new HashMap<>();
        
        for(int i = 0; i<s.length(); i++){
            if(map.containsKey(s.charAt(i))){
                start = Math.max(start, map.get(s.charAt(i))+1);
            }
            
            map.put(s.charAt(i), i);
            
            maxL = Math.max(maxL, i-start+1);
        }
        
        return maxL;
    }
}

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        
        vector<int> map(128, -1);
        int start = 0;
        int maxLen = 0;
        for(int i = 0; i<s.size(); i++){
            if(map[s[i]] != -1){
                start = max(start, map[s[i]]+1);
            }
            map[s[i]] = i;
            maxLen = max(maxLen, i-start+1);
        }
        
        return maxLen;
    }
};